add nubOrd and friends

gbaz opened the pull request:
resolves #439

I think I have the asymptotics for `nubInt` right, but a sanity check wouldn't hurt...

treeowl commented:
Could you add QuickCheck properties?

On Jan 25, 2018 4:59 PM, "gbaz" <notifications@github.com> wrote:

> resolves #439 <https://github.com/haskell/containers/issues/439>
>
> I think I have the asymptotics for nubInt right, but a sanity check
> wouldn't hurt...
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/haskell/containers/pull/515
> Commit Summary
>
>    - add ordnub and friends
>
> File Changes
>
>    - *A* Data/Containers/ListUtils.hs
>    <https://github.com/haskell/containers/pull/515/files#diff-0> (65)
>    - *M* containers.cabal
>    <https://github.com/haskell/containers/pull/515/files#diff-1> (1)
>
> Patch Links:
>
>    - https://github.com/haskell/containers/pull/515.patch
>    - https://github.com/haskell/containers/pull/515.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/containers/pull/515>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_Y9w0DYRVp-ipEmaciQB7QoyUEqtks5tOPkvgaJpZM4RtjZB>
> .
>

treeowl commented on Data/Containers/ListUtils.hs:
```diff
@@ -0,0 +1,65 @@
+    go s (x:xs) = if x `Set.member` s then go s xs
+                                      else x : go (Set.insert x s) xs
+
+-- | The `nubOrdOn` function behaves just like `nubOrd` except it preforms comparisons not on the
```
Performs, not preforms.

treeowl commented on Data/Containers/ListUtils.hs:
```diff
@@ -0,0 +1,65 @@
+                                         else x : go (IntSet.insert x s) xs
+
+-- | The `nubIntOn` function behaves just like 'nubInt' except it preforms comparisons not on the
+-- original datatype, but a user-specified projection from that datatype to 'Int'.
```
Performs, not preforms.

treeowl commented on Data/Containers/ListUtils.hs:
```diff
@@ -0,0 +1,65 @@
+-- | The `nubIntOn` function behaves just like 'nubInt' except it preforms comparisons not on the
+-- original datatype, but a user-specified projection from that datatype to 'Int'.
+nubIntOn :: (a -> Int) -> [a] -> [a]
+nubIntOn f l = go IntSet.empty l
```
Eta reduce. And it's probably slightly faster to make `f` an argument to `go`, here and above.

treeowl commented:
Actually, I changed my mind. We want to let `f` have the chance to inline, so keep it pulled out.

treeowl commented:
However however, inlining `f` is only useful if we make this participate in list fusion. So let's do that too.

gbaz commented:
Added tests, fixed typo, eta reduced.

I did not add fusion. I took a run at it and it wasn't straightforward. Doing a foldr doesn't really work because it reverses the direction in which things are dropped. So that means making it a good consumer is hard. Further, the way in which we end up building out of the fold (which involves taking a snd of the result, since we need to pass around a pair) is rather ugly and I think gets in the way of making it a good producer.

treeowl commented:
You can definitely make it a good consumer nicely.

nubOrd xs = foldr go stop xs S.empty

On Jan 26, 2018 1:28 AM, "gbaz" <notifications@github.com> wrote:

> Added tests, fixed typo, eta reduced.
>
> I did not add fusion. I took a run at it and it wasn't straightforward.
> Doing a foldr doesn't really work because it reverses the direction in
> which things are dropped. So that means making it a good consumer is hard.
> Further, the way in which we end up building out of the fold (which
> involves taking a snd of the result, since we need to pass around a pair)
> is rather ugly and I think gets in the way of making it a good producer.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/containers/pull/515#issuecomment-360695643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABzi_Rdx12RCeiDDrd-_uulbFEWrw1Vdks5tOXCIgaJpZM4RtjZB>
> .
>

gbaz commented:
I don't know what that means. In any case, I tried, and didn't get it working. If you know how to get it working, I suggest we merge first, and then you make a further patch.

treeowl commented:
Sure.

On Jan 26, 2018 1:46 AM, "gbaz" <notifications@github.com> wrote:

I don't know what that means. In any case, I tried, and didn't get it
working. If you know how to get it working, I suggest we merge first, and
then you make a further patch.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/pull/515#issuecomment-360697951>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABzi_fEObtSQzcDG-2j09S4MPKfXMb5iks5tOXTPgaJpZM4RtjZB>
.

gbaz commented:
Ok, is this good to merge then?

treeowl commented:
Thanks.
